### PR TITLE
DataFrame metadata cleanup

### LINKED ISSUE #1209: Dask.dataframe metadata handling loses information
Currently metadata is passed into dataframe objects as either a list of columns/column name, or a pandas object. If it's a pandas object, the dtype is said to be known, and the object is stored as a `_pd` attribute. Otherwise, the dtype is unknown, and a dummy pandas object with all object dtypes is created. Using the scheme, metadata is bubbled throughout, and most things work. However, this fails in the following situation:

1. Create a dask dataframe with column input, and unknown dtype. This means that: `_pd` is a dataframe with all object dtype, and `_known_dtype` is False.
2. Do any operation that doesn't first compute the dtype if it's unknown (for example, extract a column). In this case, the pandas dataframe would be converted to a pandas series (still with object dtype), and passed in as metadata to the `dd.Series` object.
3. Because `dd.Series` received a pandas object instead of a string/list, it sets `_known_dtype = True`, which causes issues later.

Example in code:

```python
In [1]: import dask.dataframe as dd

In [2]: import pandas as pd

In [3]: df = pd.DataFrame({'x': [1, 2, 2, 2, 4], 'y': ['a', 'b', 'b', 'b', 'c']})

In [4]: dsk = {('test', 0): df}

In [5]: ddf = dd.DataFrame(dsk, 'test', ['x', 'y'], [0, 4])

In [6]: ddf._known_dtype    # dtype is unknown here
Out[6]: False

In [7]: ddf._pd
Out[7]:
Empty DataFrame
Columns: [x, y]
Index: []

In [8]: x = ddf.x

In [9]: x._known_dtype    # But after selecting a column, `_known_dtype` is incorrectly True
Out[9]: True

In [10]: x._pd
Out[10]: Series([], Name: x, dtype: object)

In [11]: x.dtype    # wrong result :/
Out[11]: dtype('O')

In [12]: x.compute().dtype
Out[12]: dtype('int64')
```

Proposal:

- Track state as a tuple (or namedtuple) of (pandas object, known_dtype). This will be stored as a `_meta` attribute. We could forward the `_pd` and `_known_dtype` attributes as properties to just index this tuple. I like the use of a tuple, as it ties the two things together, since both are part of the state of the object.
- Constructors can also accept a tuple of this form, and will set the state appropriately
- Rigorous testing to track down and squash all the places where we're being too lose with dtype inference (this is the non-fun part :/)

### PR DESCRIPTION
**Pull request by jcrist**
This PR is an attempt to enforce that dask dataframes always know their full metadata (fixes #1209). Previously we allowed partial knowledge, with full dtype/column name computation done lazily if needed. This seemed like a good idea at the time, but masked many hidden metadata inference issues (functions could forget to pass along dtype, and we'd just recompute later). We also had to keep track of which bits of metadata were known, and which were unknown, resulting in frequent incorrect dtype inference.

Inference with pandas is much trickier than it is with numpy, as the dtype of the output series can change depending on values as well as input dtypes. Some functions also return different dtypes if a series is empty vs if it isn't. To get around this, a `_pd_nonempty` attribute is added to all `_Frame` objects, which matches `_pd`'s metadata, but contains one row of fake data. This helps for many situations, but not all.

Todo:
- [x] Figure out a way to smoothly deprecate the old behavior (unless we decide to make this a breaking change)
- [x] Figure out a good way to handle methods that take user provided functions like `apply`, `map_partitions`, etc...
- [x] Improve test suite to check that metadata actually matches on outputs.

### CODE REVIEW

**Comment by nirizr** [comment]
Follow + volunteer to contribute.

Right now I am manually creating dask dataframes and enjoy dask's ability to keep going without full knowledge prior to computation.
Therefore this interests me both as a "test subject" and as a way to make internal adjustments before the change.

File: dask/dataframe/categorical.py [hunk 1]
```diff
@@ -27,15 +27,17 @@ def categorize(df, columns=None, **kwargs):
...
         columns = [columns]
 
     distincts = [df[col].drop_duplicates() for col in columns]
     values = compute(*distincts, **kwargs)
 
     func = partial(_categorize_block, categories=dict(zip(columns, values)))
-    return df.map_partitions(func, columns=df.columns)
+
+    meta = func(df._pd)
+    return df.map_partitions(func, columns=meta)
```
Comment (mrocklin) [inline dask/dataframe/categorical.py]: Request that we eventually change `columns=` to `meta=` or some other word like empty, sample, dummy, etc..

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -190,28 +190,30 @@ def _build_pd(cls, metadata):
...
-                known_dtype = False
-        return _pd, known_dtype
+            _pd = cls._partition_type(columns=metadata)
+            # known_dtype = False
+        return _pd
+
+    @property
+    def _pd_nonempty(self):
+        """ A non-empty version of `_pd` with fake data."""
+        return nonempty_pd(self._pd)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Request that we change the use of `_pd` to `_meta` or `_sample` etc..

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -1480,27 +1504,32 @@ class DataFrame(_Frame):
...
+        An empty dataframe with appropriate column names and dtype.
     divisions: tuple of index values
         Values along which we partition our blocks on the index
     """
 
     _partition_type = pd.DataFrame
     _token_prefix = 'dataframe-'
 
-    def __new__(cls, dsk, name, columns, divisions):
+    def __new__(cls, dsk, name, meta, divisions):
```
Comment (mrocklin) [inline dask/dataframe/core.py]: ah, I see that you're doing this renaming, great.

File: dask/dataframe/tests/test_dataframe.py [hunk 4]
```diff
@@ -963,16 +963,18 @@ def test_append2():
     dsk = {('x', 0): pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]}),
            ('x', 1): pd.DataFrame({'a': [4, 5, 6], 'b': [3, 2, 1]}),
            ('x', 2): pd.DataFrame({'a': [7, 8, 9], 'b': [0, 0, 0]})}
-    ddf1 = dd.DataFrame(dsk, 'x', ['a', 'b'], [None, None])
+    meta = make_empty_frame({'a': 'i8', 'b': 'i8'})
+    ddf1 = dd.DataFrame(dsk, 'x', meta, [None, None])
```
Comment (mrocklin) [inline dask/dataframe/tests/test_dataframe.py]: In some cases it might make sense to start creating dataframe objects with `dd.from_pandas` instead of manually by hand.  This style of test is mostly a holdover from before `from_pandas`.

**Comment by mrocklin** [comment]
> Figure out a good way to handle methods that take user provided functions like apply, map_partitions, etc...

Try the function on `_nonempty_pd` and if it works great, if it doesn't try on `_pd` and if that doesn't work then raise an error?

**Comment by mrocklin** [comment]
> Improve test suite to check that metadata actually matches on outputs.

Hooray `eq`!

**Comment by jcrist** [comment]
This is proving more difficult than expected. The main source of frustration is the implicit conversion of integers to floats if an operation finds missing data. For example, joining two indexed dataframes together with integer columns may result in integer columns or float columns, depending on if any missing rows are found in the join. This is impossible to infer beforehand, and is also impossible to infer by just evaluating the first few rows of a partition (as is currently done) because it depends on a global presence of missingness in the column.

I'm not sure what to do here. Perhaps just do our best, but don't necessarily error in tests if `float64` != `int64` for operations where missingness can appear?

**Comment by mrocklin** [comment]
That sounds reasonable to me.  The only checks that really seem to care about this seem to be checks for number-y-ness.  Alternatively we could coerce everything to float if it came to that?

**Comment by jcrist** [comment]
This is ready for review. A few notes:

- In some places there are breaking changes to the api. These are limited to only non-pandas replicating functions (e.g. `map_partitions`, `elemwise`, `apply_concat_apply`), as I feel users making use of these will be more willing to update their code (also, probably have fewer users here).
- All constructors now will error if metadata isn't provided on construction. For most users this won't be a problem, but it's also currently a breaking change. I'm apt to punt on this and not deprecate smoothly, but I could be persuaded to deprecate nicely if needed.
- All places that take metadata can take either:
     - A pandas object
     - A dict of `{name: dtype}` (from which a dataframe will be constructed)
     - A tuple of `(name, dtype)` (from which a series will be constructed)
 This was the best "easy" api I could come up with, and I'm pretty happy with it. If you have a better idea, I'd love to hear it.
- I know I still need to up coverage in a few places. Waiting for feedback/not a friday evening :/.

**Comment by mrocklin** [comment]
In the dict case how do we determine the order of the columns?

**Comment by jcrist** [comment]
That's a good point, I suppose we don't. This hasn't come up yet, as all cases where I've used the `dict` case, I've also created the dataframe with a `dict`, so pandas default ordering worked. Perhaps a list of tuples instead of a dict?

**Comment by mrocklin** [comment]
Or ordered dict, or maybe it doesn't matter at all for how we use it?  I'm not sure we have operations that really depend on the ordering of columns within the dataframe.  Presumably if the test suite passes (which I believe checks for this) then we don't ever actually use it.

**Comment by mrocklin** [comment]
Also, please accept a bit "thank you" for diving into this.

File: dask/dataframe/tests/test_io.py [hunk 5]
```diff
@@ -523,11 +512,10 @@ def test_Series_from_dask_array():
 def test_from_dask_array_compat_numpy_array():
     x = da.ones((3, 3, 3), chunks=2)
 
-    msg = r"from_array does not input more than 2D array, got array with shape \(3, 3, 3\)"
-    with tm.assertRaisesRegexp(ValueError, msg):
+    with pytest.raises(ValueError):
```
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: Why this change?

File: dask/dataframe/utils.py [hunk 6]
```diff
@@ -167,6 +230,8 @@ def _check_dask(dsk, check_names=True):
             assert isinstance(dsk._pd, pd.Index), type(dsk._pd)
             if check_names:
                 assert dsk._pd.name == result.name
+            if check_dtypes:
+                assert_dask_dtypes(dsk, result)
```
Comment (mrocklin) [inline dask/dataframe/utils.py]: Happy to see more checks going into `eq`

**Comment by mrocklin** [comment]
At a first glance through this all looks pretty reasonable to me.  I haven't checked things like coverage though.

File: dask/dataframe/tests/test_io.py (hunk 5 above)
Comment (jcrist) [inline dask/dataframe/tests/test_io.py]: I had to change how that function was implemented, and pulled the error checking into our code (which changed what they read). I didn't see the value in checking that the messages matched, but could change back if desired.

**Comment by jcrist** [comment]
Ok, I think this is good to go now (minus a good review). I expect there might be some unexpected fallout from this, so it'd be good to get it in master soon so others can start playing with it and finding the edge cases that I couldn't. A few breaking changes I'd like to point out:

- `df.map_partitions` signature is changed without a deprecation (there wasn't an easy way to change it in a backwards compatible way).

  ```python
# old signature
df.map_partitions(self, func, columns=no_default, *args, **kwargs)
# new signature
df.map_partitions(self, func, *args, **kwargs)    # meta is in `kwargs`
  ```
  The reason for this change was that the old signature made it impossible to pass in `args` without providing something to `columns`. Coupled with the desire to rename the `columns` keyword, and the fact that many internal uses of `map_partitions` passed in `columns` as a positional, it was difficult to find a way to smoothly deprecate. It should be noted that the function `dd.map_partitions` only has a positional renamed, and remains the same otherwise. I could be convinced to revert this change if needed.

- All object constructors (i.e. `dd.DataFrame`) will now error if not fully typed. This shouldn't affect most users, but is a breaking change for those mucking around in dataframe internals. I could probably find a way to smoothly deprecate that, but I'd rather not.

- `_pd` and `_pd_nonempty` have been renamed to `_meta` and `_meta_nonempty`. No deprecation of the old attributes, as they were internal api. Could deprecate easily if desired.

- Several positional arguments have been renamed, mostly in the internal api. This should only affect users that were using those internals, and passing in positionals as keywords. I doubt this is common, but it is a breaking change.

Everything else should be a smooth transition. Functions like `map` and `apply` have the `columns` keyword deprecated in favor of `meta`.

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -283,48 +263,56 @@ def cache(self, cache=Cache):
...
             Function applied to each blocks
-        columns : tuple or scalar
-            Column names or name of the output. Defaults to names of data itself.
-            When tuple is passed, DataFrame is returned. When scalar is passed,
-            Series is returned.
+        args, kwargs :
+            Arguments and keywords to pass to the function. The partition will
+            be the first argument, and these will be passed *after*.
+        meta : pd.DataFrame, pd.Series, dict, tuple, optional
+            Metadata describing the output DataFrame or Series. Should have
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Suggest something like the following:

>  An empty pandas DataFrame or Series that matches the dtypes and column names of the output.  Alternatively ...

Leading with the term metadata might be confusing to newcomers.

File: dask/dataframe/core.py [hunk 8]
```diff
@@ -1087,8 +1087,8 @@ class Series(_Frame):
     _name: str
         The key prefix that specifies which keys in the dask comprise this
         particular Series
-    name: scalar or None
-        Series name.  This metadata aids usability
+    meta: pd.Series
+        An empty Series with appropriate name and dtype.
```
Comment (mrocklin) [inline dask/dataframe/core.py]: define appropriate.  Perhaps "An empty Pandas Series with name and type matching the expected output" (if that fits)

File: dask/dataframe/core.py [hunk 9]
```diff
@@ -2245,14 +2253,15 @@ def apply_concat_apply(args, chunk=None, aggregate=None,
     else:
         dsk2 = {(b, 0): (apply, aggregate, [conc], aggregate_kwargs)}
 
-    if columns is no_default:
+    if meta is no_default:
+        # TODO: handle this!
```
Comment (mrocklin) [inline dask/dataframe/core.py]: What is the reason for this TODO

File: dask/dataframe/core.py [hunk 10]
```diff
@@ -2292,50 +2301,60 @@ def _get_return_type(arg, metadata):
...
     elif isinstance(x, dict):
         res = {}
         for k in x:
-            res[k] = _extract_pd(x[k])
+            res[k] = _extract_meta(x[k], nonempty)
         return res
+    elif isinstance(x, Scalar):
+        # TODO: clean this up. This is fine for most things,
+        # but will fail if the scalar isn't numeric.
+        return 1
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Should we handle this now?  Should we raise an issue if not?

File: dask/dataframe/core.py [hunk 11]
```diff
@@ -2292,50 +2301,60 @@ def _get_return_type(arg, metadata):
...
     Apply a function using args / kwargs. If arguments contain dd.DataFrame /
-    dd.Series, using internal cache (``_pd``) for calculation
+    dd.Series, using internal cache (``_meta``) for calculation
     """
-    return func(*_extract_pd(args), **_extract_pd(kwargs))
+    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
 
 
-def map_partitions(func, metadata, *args, **kwargs):
+def map_partitions(func, meta, *args, **kwargs):
```
Comment (mrocklin) [inline dask/dataframe/core.py]: While changing the `map_partitions` method API should we change this one as well?

**Comment by mrocklin** [comment]
`_pd` was sufficiently widespread that I think it would be good to deprecate it with a property, even though it was internal.

File: dask/dataframe/core.py (hunk 10 above)
Comment (jcrist) [inline dask/dataframe/core.py]: This is coupled with the fact that `Scalar` doesn't keep any metadata about it (see this old `TODO`: https://github.com/dask/dask/blob/master/dask/dataframe/core.py#L70). I planned to punt on this for now (it's no worse than what currently exists), and fix it in a later PR. Could raise an issue I suppose.

File: dask/dataframe/core.py (hunk 9 above)
Comment (jcrist) [inline dask/dataframe/core.py]: To remind me to fix it before merging :). Thanks, missed that one.

File: dask/dataframe/core.py (hunk 11 above)
Comment (jcrist) [inline dask/dataframe/core.py]: Fixed.

File: dask/dataframe/core.py (hunk 9 above)
Comment (jcrist) [inline dask/dataframe/core.py]: Fixed.

File: dask/dataframe/core.py (hunk 8 above)
Comment (jcrist) [inline dask/dataframe/core.py]: Fixed.

File: dask/dataframe/core.py (hunk 7 above)
Comment (jcrist) [inline dask/dataframe/core.py]: Fixed.

**Comment by jcrist** [comment]
Anything else here? Planning to merge later today if no comment.

**Comment by mrocklin** [comment]
Nothing from me.  +1 to merge
